addons/sale_stock/models/sale_order.py

```diff
@@ -190,6 +191,57 @@ class SaleOrderLine(models.Model):
     product_packaging = fields.Many2one('product.packaging', string='Package', default=False)
     route_id = fields.Many2one('stock.location.route', string='Route', domain=[('sale_selectable', '=', True)], ondelete='restrict')
     move_ids = fields.One2many('stock.move', 'sale_line_id', string='Stock Moves')
+    is_storable = fields.Boolean(compute='_compute_is_storable')
+    qty_at_date = fields.Float(compute='_compute_qty_at_date')
+    scheduled_date = fields.Datetime(compute='_compute_qty_at_date')
+    qty_available = fields.Float(compute='_compute_qty_at_date')
+    qty_on_hand = fields.Float(compute='_compute_qty_at_date')
+    warehouse_view_id = fields.Many2one(related='order_id.warehouse_id.view_location_id')
+    qty_to_deliver = fields.Float(compute='_compute_qty_to_deliver')
+
+    @api.depends('product_uom_qty', 'qty_delivered')
+    def _compute_qty_to_deliver(self):
+        """ technical fields to hide or display the 'inventory on hand widget' on
+        sale order lines"""
+        for line in self:
+            line.qty_to_deliver = line.product_uom_qty - line.qty_delivered
+
+    @api.depends('customer_lead', 'product_uom_qty', 'order_id.warehouse_id', 'order_id.commitment_date')
+    def _compute_qty_at_date(self):
+        """ Compute the quantity forecasted of product at delivery date. There are
+        two cases:
+         1. The quotation has a commitment_date, we take it as delivery date
+         2. The quotation hasn't commitment_date, we compute the estimated delivery
+            date based on lead time"""
+        duplicate_product = defaultdict(lambda: 0)
+        grouped_lines = defaultdict(self.env['sale.order.line'].browse)
+        for line in self:
+            # Group lines by order_id.warehouse and delivery date
+            if line.product_uom_qty - line.qty_delivered <= 0 or not line.product_id:
+                continue
+            warehouse = line.order_id.warehouse_id
+            if line.order_id.commitment_date:
+                date = line.order_id.commitment_date
+            else:
+                confirm_date = line.order_id.confirmation_date if line.order_id.state in ['sale', 'done'] else datetime.now()
+                date = confirm_date + timedelta(days=line.customer_lead or 0.0)
+            grouped_lines[(warehouse.id, date)] |= line
+        for data, lines in grouped_lines.items():
+            # Read quantities in batch depending on the warehouse and to_date
+            res = {product['id']: (product['qty_available'], product['free_qty'], product['virtual_available']) for product in lines.mapped('product_id').with_context(to_date=data[1], warehouse=data[0]).read([
+                'id',
+                'qty_available',
+                'free_qty',
+                'virtual_available',
+            ])}
+            for line in lines:
+                line.scheduled_date = data[1]
+                line.qty_on_hand, line.qty_available, line.qty_at_date = res[line.product_id.id]
```

Comment:
qty_on_hand, qty_available, qty_at_date = qties_per_product[line.product_id.id]
line.qty_on_hand = qty_on_hand - duplicate_product[line.product_id.id]
...
duplicate_product should also be renamed to "qty_processed_per_product" aybe